File: go/systemparams/machine.go
package main

import (
	"fmt"

	"github.com/frame-lang/frame-demos/systemparams/framelang"
)

func NewSystemParams(stateMsg string, enterMsg string) SystemParams {
	m := &systemParamsStruct{}

	// Validate interfaces
	var _ SystemParams = m
	var _ SystemParams_actions = m
	m._compartment_ = NewSystemParamsCompartment(SystemParamsState_Begin)
	m._compartment_.StateArgs["stateMsg"] = stateMsg

	// Initialize domain

	// Send system start event
	params := make(map[string]interface{})
	params["enterMsg"] = enterMsg
	e := framelang.FrameEvent{Msg: ">", Params: params}
	m._mux_(&e)
	return m
}

type SystemParamsState uint

const (
	SystemParamsState_Begin SystemParamsState = iota
)

type SystemParams interface {
}

type SystemParams_actions interface {
	print(msg string)
}

type systemParamsStruct struct {
	_compartment_     *SystemParamsCompartment
	_nextCompartment_ *SystemParamsCompartment
}

//====================== Multiplexer ====================//

func (m *systemParamsStruct) _mux_(e *framelang.FrameEvent) {
	switch m._compartment_.State {
	case SystemParamsState_Begin:
		m._SystemParamsState_Begin_(e)
	}

	if m._nextCompartment_ != nil {
		nextCompartment := m._nextCompartment_
		m._nextCompartment_ = nil
		if nextCompartment._forwardEvent_ != nil &&
			nextCompartment._forwardEvent_.Msg == ">" {
			m._mux_(&framelang.FrameEvent{Msg: "<", Params: m._compartment_.ExitArgs, Ret: nil})
			m._compartment_ = nextCompartment
			m._mux_(nextCompartment._forwardEvent_)
		} else {
			m._do_transition_(nextCompartment)
			if nextCompartment._forwardEvent_ != nil {
				m._mux_(nextCompartment._forwardEvent_)
			}
		}
		nextCompartment._forwardEvent_ = nil
	}
}

//===================== Machine Block ===================//

func (m *systemParamsStruct) _SystemParamsState_Begin_(e *framelang.FrameEvent) {
	switch e.Msg {
	case ">":
		m.print((m._compartment_.StateArgs["stateMsg"].(string)) + " " + e.Params["enterMsg"].(string))
		return
	}
}

//=============== Machinery and Mechanisms ==============//

func (m *systemParamsStruct) _transition_(compartment *SystemParamsCompartment) {
	m._nextCompartment_ = compartment
}

func (m *systemParamsStruct) _do_transition_(nextCompartment *SystemParamsCompartment) {
	m._mux_(&framelang.FrameEvent{Msg: "<", Params: m._compartment_.ExitArgs, Ret: nil})
	m._compartment_ = nextCompartment
	m._mux_(&framelang.FrameEvent{Msg: ">", Params: m._compartment_.EnterArgs, Ret: nil})
}

//===================== Actions Block ===================//

func (m *systemParamsStruct) print(msg string) {

	fmt.Println(msg)

}

/********************************************************

// Unimplemented Actions


********************************************************/

//=============== Compartment ==============//

type SystemParamsCompartment struct {
	State          SystemParamsState
	StateArgs      map[string]interface{}
	StateVars      map[string]interface{}
	EnterArgs      map[string]interface{}
	ExitArgs       map[string]interface{}
	_forwardEvent_ *framelang.FrameEvent
}

func NewSystemParamsCompartment(state SystemParamsState) *SystemParamsCompartment {
	c := &SystemParamsCompartment{State: state}
	c.StateArgs = make(map[string]interface{})
	c.StateVars = make(map[string]interface{})
	c.EnterArgs = make(map[string]interface{})
	c.ExitArgs = make(map[string]interface{})
	return c
}

File: javaframedemos/systemparams/machine.java
import java.util.*;

// emitted from framec_v0.10.0
// get include files at https://github.com/frame-lang/frame-ancillary-files

class SystemParams {
    
    private SystemParamsCompartment _compartment_;
    private SystemParamsCompartment _nextCompartment_;
    
    SystemParams(String stateMsg, String enterMsg) {
        
        // Create and intialize start state compartment.
        this._compartment_ = new SystemParamsCompartment(this._state_);
        this._nextCompartment_ = null;
        this._compartment_.stateArgs.put("stateMsg", stateMsg);
        this._compartment_.enterArgs.put("enterMsg", enterMsg);
        
        
        // Send system start event
        FrameEvent frame_event = new FrameEvent(">", this._compartment_.enterArgs);
        this._mux_(frame_event);
    
    }
    
    // states enum
    private enum SystemParamsState {
        
        BEGIN(0);
        
        public final int value;
        
        private SystemParamsState(int value) {
            this.value=value;
        }
        
        public int getValue() {
            return value;
        }
    }
    
    //====================== Multiplexer ====================//
    
    private void _mux_(FrameEvent e) {
        if(this._compartment_.state == SystemParamsState.BEGIN.getValue()) {
            this._sBegin_(e);
        }
        
        
        if(this._nextCompartment_ != null) {
            SystemParamsCompartment nextCompartment = this._nextCompartment_;
            this._nextCompartment_ = null;
            if(nextCompartment._forwardEvent != null && 
             nextCompartment._forwardEvent._message == ">") {
                this._mux_(new FrameEvent( "<", this._compartment_.exitArgs));
                this._compartment_ = nextCompartment;
                this._mux_(nextCompartment._forwardEvent);
            } else {
                this._doTransition_(nextCompartment);
                if(nextCompartment._forwardEvent != null) {
                    this._mux_(nextCompartment._forwardEvent);
                }
            }
            nextCompartment._forwardEvent = null;
        }
    }
    
    
    //===================== Machine Block ===================//
    
    private void _sBegin_(FrameEvent e) {
        if(e._message == ">") {
            print_do(((String) this._compartment_.stateArgs.get("stateMsg")) + " " + String.valueOf(((String) e._parameters.get("enterMsg"))));
            return;
        }
    }
    
    //===================== Actions Block ===================//
    
    protected void print_do(String msg) {
        
        System.out.println(msg);
    
    }
    
    // Unimplemented Actions
    
    
    //=============== Machinery and Mechanisms ==============//
    
    private int _state_;
    
    private void _transition_(SystemParamsCompartment compartment) {
        this._nextCompartment_ = compartment;
    }
    
    private void _doTransition_(SystemParamsCompartment nextCompartment) {
        this._mux_(new FrameEvent("<", this._compartment_.exitArgs));
        this._compartment_ = nextCompartment;
        this._mux_(new FrameEvent(">", this._compartment_.enterArgs));
    }
    
}

//=============== Compartment ==============//

class SystemParamsCompartment {

    int state;
    
    SystemParamsCompartment(int state) {
        this.state = state;
    }
    
    HashMap<String, Object> stateArgs = new HashMap<String, Object>();
    HashMap<String, Object> stateVars = new HashMap<String, Object>();
    HashMap<String, Object> enterArgs = new HashMap<String, Object>();
    HashMap<String, Object> exitArgs = new HashMap<String, Object>();
    FrameEvent _forwardEvent = new FrameEvent();
}


class SystemParamsController extends SystemParams {
	SystemParamsController(String stateMsg, String enterMsg) {
	  super(stateMsg, enterMsg);
	}
}
